In [1]:
%pylab inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
plt.rc('figure', figsize=(10, 6))
plt.style.use('ggplot')

/Users/lynngao/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py:273: UserWarning: Matplotlib is building the font cache using fc-list. This may take a moment.
  warnings.warn('Matplotlib is building the font cache using fc-list. This may take a moment.')


Populating the interactive namespace from numpy and matplotlib


### Load data

In [2]:
df = pd.read_csv("../data/NYPD_Complaint_Data_Historic.csv")
#df = pd.read_csv("../data/NYPD_Complaint_Data_Historic.csv", index_col='CMPLNT_NUM')

/Users/lynngao/anaconda/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
df.head()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM      RPT_DT  \
0   101109527   12/31/2015     23:45:00          NaN          NaN  12/31/2015   
1   153401121   12/31/2015     23:36:00          NaN          NaN  12/31/2015   
2   569369778   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
3   968417082   12/31/2015     23:30:00          NaN          NaN  12/31/2015   
4   641637920   12/31/2015     23:25:00   12/31/2015     23:30:00  12/31/2015   

   KY_CD                        OFNS_DESC  PD_CD  \
0    113                          FORGERY  729.0   
1    101  MURDER & NON-NEGL. MANSLAUGHTER    NaN   
2    117                  DANGEROUS DRUGS  503.0   
3    344     ASSAULT 3 & RELATED OFFENSES  101.0   
4    344     ASSAULT 3 & RELATED OFFENSES  101.0   

                          PD_DESC              ...               ADDR_PCT_CD  \
0  FORGERY,ETC.,UNCLASSIFIED-FELO              ...                      44.0   
1                             NaN              ...                     103.0   
2  CONTROLLED SUBSTANCE,INTENT TO              ...                      28.0   
3                       ASSAULT 3              ...                     105.0   
4                       ASSAULT 3              ...                      13.0   

  LOC_OF_OCCUR_DESC    PREM_TYP_DESC PARKS_NM  HADEVELOPT X_COORD_CD  \
0            INSIDE   BAR/NIGHT CLUB      NaN         NaN  1007314.0   
1           OUTSIDE              NaN      NaN         NaN  1043991.0   
2               NaN            OTHER      NaN         NaN   999463.0   
3            INSIDE  RESIDENCE-HOUSE      NaN         NaN  1060183.0   
4          FRONT OF            OTHER      NaN         NaN   987606.0   

  Y_COORD_CD   Latitude  Longitude                        Lat_Lon  
0   241257.0  40.828848 -73.916661  (40.828848333, -73.916661142)  
1   193406.0  40.697338 -73.784557  (40.697338138, -73.784556739)  
2   231690.0  40.802607 -73.945052  (40.802606608, -73.945051911)  
3   177862.0  40.654549 -73.726339  (40.654549444, -73.726338791)  
4   208148.0  40.738002 -73.987891     (40.7380024, -73.98789129)  

[5 rows x 24 columns]

In [13]:
df.tail()

CMPLNT_NUM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT CMPLNT_TO_TM  \
5101226   837109942   01/01/2005     00:45:00          NaN          NaN   
5101227   147823150   01/01/2005     00:30:00          NaN          NaN   
5101228   832021762   01/01/2003     17:00:00   01/01/2006     20:00:00   
5101229   395395139   05/01/1999     08:00:00   12/21/2005     08:00:00   
5101230   673950804          NaN     22:00:00   01/01/2006     13:00:00   

             RPT_DT  KY_CD                       OFNS_DESC  PD_CD  \
5101226  01/01/2006    351  CRIMINAL MISCHIEF & RELATED OF  259.0   
5101227  01/01/2006    344    ASSAULT 3 & RELATED OFFENSES  101.0   
5101228  01/01/2006    351  CRIMINAL MISCHIEF & RELATED OF  254.0   
5101229  01/01/2006    109                   GRAND LARCENY  405.0   
5101230  01/01/2006    109                   GRAND LARCENY  457.0   

                                                   PD_DESC  \
5101226                   CRIMINAL MISCHIEF,UNCLASSIFIED 4   
5101227                                          ASSAULT 3   
5101228                     MISCHIEF, CRIMINAL 4, OF MOTOR   
5101229              LARCENY,GRAND BY THEFT OF CREDIT CARD   
5101230  LARCENY,GRAND OF VEHICULAR/MOTORCYCLE ACCESSORIES   

                     ...               ADDR_PCT_CD LOC_OF_OCCUR_DESC  \
5101226              ...                      42.0            INSIDE   
5101227              ...                      33.0            INSIDE   
5101228              ...                      25.0            INSIDE   
5101229              ...                     114.0            INSIDE   
5101230              ...                      61.0          FRONT OF   

                      PREM_TYP_DESC PARKS_NM  HADEVELOPT X_COORD_CD  \
5101226  RESIDENCE - PUBLIC HOUSING      NaN         NaN  1011978.0   
5101227      RESIDENCE - APT. HOUSE      NaN         NaN  1001995.0   
5101228                      STREET      NaN         NaN  1002076.0   
5101229             RESIDENCE-HOUSE      NaN         NaN  1004663.0   
5101230                      STREET      NaN         NaN   999199.0   

        Y_COORD_CD   Latitude  Longitude                        Lat_Lon  
5101226   240173.0  40.825860 -73.899813  (40.825859654, -73.899812566)  
5101227   246271.0  40.842623 -73.935868  (40.842622599, -73.935867743)  
5101228   231440.0  40.801916 -73.935614   (40.801915542, -73.93561439)  
5101229   216473.0  40.760830 -73.926316  (40.760829504, -73.926315689)  
5101230   163370.0  40.615085 -73.946157  (40.615084921, -73.946157157)  

[5 rows x 24 columns]

In [15]:
df.shape

(5101231, 24)

In [4]:
df.describe()

/Users/lynngao/anaconda/lib/python2.7/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


CMPLNT_NUM         KY_CD         PD_CD   ADDR_PCT_CD    X_COORD_CD  \
count  5.101231e+06  5.101231e+06  5.096657e+06  5.100841e+06  4.913085e+06   
mean   5.499733e+08  2.932191e+02  4.179772e+02  6.346322e+01  1.004610e+06   
std    2.597636e+08  1.480821e+02  2.179312e+02  3.444424e+01  2.144742e+04   
min    1.000002e+08  1.010000e+02  1.010000e+02  1.000000e+00  9.133190e+05   
25%    3.250451e+08  1.170000e+02           NaN           NaN           NaN   
50%    5.499458e+08  3.410000e+02           NaN           NaN           NaN   
75%    7.748865e+08  3.510000e+02           NaN           NaN           NaN   
max    9.999999e+08  8.810000e+02  9.750000e+02  1.230000e+02  1.067298e+06   

         Y_COORD_CD      Latitude     Longitude  
count  4.913085e+06  4.913085e+06  4.913085e+06  
mean   2.068627e+05  4.073443e+01 -7.392651e+01  
std    3.047765e+04  8.365685e-02  7.734681e-02  
min    1.208290e+05  4.049806e+01 -7.425508e+01  
25%             NaN           NaN           NaN  
50%             NaN           NaN           NaN  
75%             NaN           NaN           NaN  
max    2.718200e+05  4.091272e+01 -7.370032e+01

## 1. Check missing data by column

In [5]:
def count_missing_value_by_col(dataframe):
    missing_count = dataframe.isnull().sum()
    missing_value_df = pd.concat([missing_count, missing_count/dataframe.shape[0] * 100], axis=1)
    missing_value_df.columns = ['row_count', 'percent']
    return missing_value_df

In [6]:
missing_summary = count_missing_value_by_col(df); missing_summary

row_count    percent
CMPLNT_NUM                 0   0.000000
CMPLNT_FR_DT             655   0.012840
CMPLNT_FR_TM              48   0.000941
CMPLNT_TO_DT         1391478  27.277298
CMPLNT_TO_TM         1387785  27.204904
RPT_DT                     0   0.000000
KY_CD                      0   0.000000
OFNS_DESC              18840   0.369323
PD_CD                   4574   0.089665
PD_DESC                 4574   0.089665
CRM_ATPT_CPTD_CD           7   0.000137
LAW_CAT_CD                 0   0.000000
JURIS_DESC                 0   0.000000
BORO_NM                  463   0.009076
ADDR_PCT_CD              390   0.007645
LOC_OF_OCCUR_DESC    1127128  22.095216
PREM_TYP_DESC          33279   0.652372
PARKS_NM             5093632  99.851036
HADEVELOPT           4848026  95.036394
X_COORD_CD            188146   3.688247
Y_COORD_CD            188146   3.688247
Latitude              188146   3.688247
Longitude             188146   3.688247
Lat_Lon               188146   3.688247

## 2. Unique value by column

In [7]:
def count_unique_value_by_col(dataframe):
    res=[]
    datatype = dataframe.dtypes
    cols = dataframe.columns
    for i in range(len(cols)):
        if datatype[i] != 'object':
            df[cols[i]].astype('str')
        res.append(len(df[cols[i]].unique()))
    res_df = pd.DataFrame(res, index=cols)
    res_df.columns = ['Count_of_unique_val']
    return res_df

In [8]:
unique_val_summary =count_unique_value_by_col(df); unique_val_summary

Count_of_unique_val
CMPLNT_NUM                     5101231
CMPLNT_FR_DT                      6372
CMPLNT_FR_TM                      1443
CMPLNT_TO_DT                      4828
CMPLNT_TO_TM                      1442
RPT_DT                            3652
KY_CD                               74
OFNS_DESC                           71
PD_CD                              416
PD_DESC                            404
CRM_ATPT_CPTD_CD                     3
LAW_CAT_CD                           3
JURIS_DESC                          25
BORO_NM                              6
ADDR_PCT_CD                         78
LOC_OF_OCCUR_DESC                    7
PREM_TYP_DESC                       71
PARKS_NM                           864
HADEVELOPT                         279
X_COORD_CD                       69533
Y_COORD_CD                       72317
Latitude                        112804
Longitude                       112808
Lat_Lon                         112827

In [9]:
# Ideally, we should have equal unique number of "KY_CD" and "OFNS_DESC."
# Check for each unique value in "KY_CD" (classfication code), 
# if it corresponds more than one unique description
for val in df['KY_CD'].unique():
    if len(df['OFNS_DESC'][df['KY_CD'] == val].value_counts()) != 1:
        print ("%s:\n%s")%(val, df['OFNS_DESC'][df['KY_CD'] == val].value_counts())

343:
OTHER OFFENSES RELATED TO THEF    9731
THEFT OF SERVICES                 2778
Name: OFNS_DESC, dtype: int64
364:
OTHER STATE LAWS (NON PENAL LA          5505
AGRICULTURE & MRKTS LAW-UNCLASSIFIED      83
OTHER STATE LAWS (NON PENAL LAW)           4
Name: OFNS_DESC, dtype: int64
120:
CHILD ABANDONMENT/NON SUPPORT    367
ENDAN WELFARE INCOMP              22
Name: OFNS_DESC, dtype: int64
124:
KIDNAPPING & RELATED OFFENSES      2300
KIDNAPPING AND RELATED OFFENSES       2
KIDNAPPING                            2
Name: OFNS_DESC, dtype: int64
345:
OFFENSES RELATED TO CHILDREN    1346
ENDAN WELFARE INCOMP             122
Name: OFNS_DESC, dtype: int64
677:
OTHER STATE LAWS                   233
NYS LAWS-UNCLASSIFIED VIOLATION      6
Name: OFNS_DESC, dtype: int64


1. For code "364", "OTHER STATE LAWS (NON PENAL LA" is apparently a typo, and can be merged with "OTHER STATE LAWS (NON PENAL LAW)".
2. For code "124", "KIDNAPPING & RELATED OFFENSES" and "KIDNAPPING AND RELATED OFFENSES" should be merged
3. We have "ENDAN WELFARE INCOMP" (Endangering the welfare of an incompetent or physically disabled person) corresponds in both classes "120" and "345".

In [10]:
df['KY_CD'][df['OFNS_DESC'] == "ENDAN WELFARE INCOMP"].value_counts()

345    122
120     22
Name: KY_CD, dtype: int64

"ENDAN WELFARE INCOMP" only exists in "120" and "345", both these categories are related to children. It may be an option to combine "120" and "345"?

In [11]:
# Check for each unique value in "OFNS_DESC" (classfication description), 
# if it corresponds more than one unique classification code
for val in df['OFNS_DESC'].unique():
    # if not a nan
    if isinstance(val, basestring):
        # if more than 1 "KY_CD" code correspond to the class description
        if len(df['KY_CD'][df['OFNS_DESC'] == val].value_counts()) != 1:
            print('--------------------------------------')
            # print the count of different codes correpond to this class description
            print ("%s:\n%s")%(val, df['KY_CD'][df['OFNS_DESC'] == val].value_counts())
            # check if for these codes, they only have this class description.
            # If yes, then the two classes can be merged.
            for code in df['KY_CD'][df['OFNS_DESC'] == val].unique():
                print(df['OFNS_DESC'][(df['KY_CD'] == code)].unique())

--------------------------------------
DANGEROUS DRUGS:
235    285790
117     62679
Name: KY_CD, dtype: int64
['DANGEROUS DRUGS' nan]
['DANGEROUS DRUGS' nan]
--------------------------------------
DANGEROUS WEAPONS:
236    73672
118    50563
Name: KY_CD, dtype: int64
['DANGEROUS WEAPONS']
['DANGEROUS WEAPONS' nan]
--------------------------------------
CRIMINAL MISCHIEF & RELATED OF:
351    433358
121     72416
Name: KY_CD, dtype: int64
['CRIMINAL MISCHIEF & RELATED OF' nan]
['CRIMINAL MISCHIEF & RELATED OF' nan]
--------------------------------------
SEX CRIMES:
233    44117
116    10853
Name: KY_CD, dtype: int64
['SEX CRIMES' nan]
['SEX CRIMES' nan]
--------------------------------------
MISCELLANEOUS PENAL LAW:
126    110468
678      8253
Name: KY_CD, dtype: int64
['MISCELLANEOUS PENAL LAW' nan]
['MISCELLANEOUS PENAL LAW' nan]
--------------------------------------
GAMBLING:
350    1901
122     122
Name: KY_CD, dtype: int64
['GAMBLING' nan]
['GAMBLING']
-----------------------------

### All of the above categories can be merged!!!!

### Top 5 unique values in each column. (This part can be revised to using Spark MapReduce.)

In [12]:
for col in df.columns:
    print(df[col].value_counts().head(5))

843057151    1
686481129    1
347469652    1
429369248    1
371124323    1
Name: CMPLNT_NUM, dtype: int64
01/01/2010    2343
01/01/2008    2273
01/01/2007    2259
01/01/2011    2253
01/01/2014    2211
Name: CMPLNT_FR_DT, dtype: int64
12:00:00    135632
15:00:00    116221
18:00:00    112102
20:00:00    106703
16:00:00    103994
Name: CMPLNT_FR_TM, dtype: int64
10/17/2014    1367
01/01/2014    1357
10/04/2013    1325
09/20/2014    1311
10/05/2012    1301
Name: CMPLNT_TO_DT, dtype: int64
12:00:00    65833
08:00:00    60224
15:00:00    59982
09:00:00    57903
16:00:00    54446
Name: CMPLNT_TO_TM, dtype: int64
05/29/2007    1903
11/01/2006    1889
10/23/2007    1854
06/21/2006    1841
06/05/2007    1829
Name: RPT_DT, dtype: int64
341    822508
578    604141
344    521611
351    433368
109    429205
Name: KY_CD, dtype: int64
PETIT LARCENY                     822498
HARRASSMENT 2                     604070
ASSAULT 3 & RELATED OFFENSES      521538
CRIMINAL MISCHIEF & RELATED OF    505774
GRAND